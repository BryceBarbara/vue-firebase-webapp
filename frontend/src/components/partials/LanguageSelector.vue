<script setup lang="tsx">
import type { DropdownOption } from 'naive-ui'
import type { FunctionalComponent } from 'vue'
import EnFlag from '~icons/flag/gb-4x3'
import DeFlag from '~icons/flag/de-4x3'

const emit = defineEmits<{
  (e: 'change', locale: string): void
}>()

const flagMap: Record<string, FunctionalComponent> = {
  en: EnFlag,
  de: DeFlag,
}

const attrs = useAttrs()
const { t, availableLocales, locale } = useI18n()

const localeOptions = computed<DropdownOption[]>(() => {
  return availableLocales.map<DropdownOption>((l) => {
    const flag = flagMap[l.toLowerCase()]
    return {
      label: t(`locale.${l}`, {}, {
        locale: 'en',
      }),
      key: l,
      icon: flag ? () => h(flag) : undefined,
      disabled: l === locale.value,
    }
  })
})
function onSelectLocale(value: string) {
  if (!availableLocales.includes(value)) {
    console.error(`Locale ${value} is not available`)
    return
  }
  locale.value = value
  emit('change', value)
}
</script>

<template>
  <n-dropdown
    trigger="hover"
    :options="localeOptions"
    @select="onSelectLocale"
  >
    <n-button v-bind="attrs" :title="t('button.change-locale')" data-test="trigger">
      <template #icon>
        <i-carbon-language />
      </template>
    </n-button>
  </n-dropdown>
</template>

<style scoped>

</style>
