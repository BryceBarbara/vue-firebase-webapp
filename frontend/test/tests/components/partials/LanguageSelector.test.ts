import { enableAutoUnmount, mount } from '@vue/test-utils'
import { afterEach, beforeAll, describe, expect, it } from 'vitest'
import { nextTick } from 'vue'
import type { I18n } from 'vue-i18n'
import LanguageSelector from '../../../../src/components/partials/LanguageSelector.vue'

enableAutoUnmount(afterEach)

const TriggerSelector = '[data-test=trigger]'

let i18n: I18n
let t = (key: string) => key

beforeAll(async () => {
  const result = await import('../../../scripts/vue-i18n.setup')
  i18n = result.i18n
  t = i18n.global.t
})

async function sleep(ms: number): Promise<void> {
  return await new Promise((resolve) => {
    setTimeout(resolve, ms)
  })
}

describe('LanguageSelector.vue', () => {
  it('should render', () => {
    const wrapper = mount(LanguageSelector)
    // Ensure there is a trigger element
    expect(wrapper.get(TriggerSelector)).toBeTruthy()
    // Ensure the trigger element has a title
    expect(wrapper.get(TriggerSelector).attributes('title')).toBe(t('button.change-locale'))
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render when expanded', async () => {
    const wrapper = mount(LanguageSelector, { attachTo: document.body })
    await wrapper.get(TriggerSelector).trigger('mouseenter')
    // Wait for the dropdown to appear
    await sleep(200)
    // Snapshot the body since the dropdown is appended to the body
    expect(document.body.innerHTML).toMatchSnapshot()
  })

  it('should be interactive', async () => {
    const wrapper = await mount(LanguageSelector, { attachTo: document.body })
    const optionBodySelector = '.n-dropdown-option-body'

    const triggerNode = wrapper.find(TriggerSelector)
    expect(triggerNode.exists()).toBe(true)

    const getOptions = () => Array.from(document.querySelectorAll(optionBodySelector))
    const isDisabledOption = (option: Element) => option.classList.contains('n-dropdown-option-body--disabled')
    const getOptionText = (option: Element) => option.querySelector('.n-dropdown-option-body__label')?.textContent
    const openDropdown = async () => {
      if (getOptions().length)
        await triggerNode.trigger('mouseleave')

      await triggerNode.trigger('mouseenter')
      // Wait for the dropdown to appear
      await sleep(200)
    }

    await openDropdown()
    const options = getOptions()
    expect(options.length).toBe(i18n.global.availableLocales.length)

    const enabledOption = options.find(option => !isDisabledOption(option))
    expect(enabledOption).toBeTruthy()
    if (enabledOption instanceof HTMLDivElement) {
      const localeToChangeTo = getOptionText(enabledOption)
      enabledOption.click()

      if (typeof i18n.global.locale !== 'string')
        // Hack to mark the computed property as dirty to force a re-evaluation
        // @ts-expect-error _dirty isn't defined
        i18n.global.locale._dirty = true

      await nextTick()
      expect(wrapper.emitted('change')?.length).toBe(1)
      await openDropdown()
      const updatedOption = getOptions().find(option => getOptionText(option) === localeToChangeTo)
      expect(updatedOption && isDisabledOption(updatedOption)).toBeTruthy()
    }
    else {
      throw new TypeError('enabledOption is not a HTMLDivElement')
    }
  })
})
