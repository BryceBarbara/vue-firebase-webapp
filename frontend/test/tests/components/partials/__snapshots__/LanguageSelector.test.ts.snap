// Vitest Snapshot v1

exports[`LanguageSelector.vue > should render 1`] = `
"<button class=\\"n-button n-button--default-type n-button--medium-type\\" tabindex=\\"0\\" type=\\"button\\" style=\\"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-bezier-ease-out: cubic-bezier(0, 0, .2, 1); --n-ripple-duration: .6s; --n-opacity-disabled: 0.5; --n-wave-opacity: 0.6; font-weight: 400; --n-color: #0000; --n-color-hover: #0000; --n-color-pressed: #0000; --n-color-focus: #0000; --n-color-disabled: #0000; --n-ripple-color: #18a058; --n-text-color: rgb(51, 54, 57); --n-text-color-hover: #36ad6a; --n-text-color-pressed: #0c7a43; --n-text-color-focus: #36ad6a; --n-text-color-disabled: rgb(51, 54, 57); --n-border: 1px solid rgb(224, 224, 230); --n-border-hover: 1px solid #36ad6a; --n-border-pressed: 1px solid #0c7a43; --n-border-focus: 1px solid #36ad6a; --n-border-disabled: 1px solid rgb(224, 224, 230); --n-width: initial; --n-height: 34px; --n-font-size: 14px; --n-padding: 0 14px; --n-icon-size: 18px; --n-icon-margin: 6px; --n-border-radius: 3px;\\" title=\\"Change language\\" data-test=\\"trigger\\">
  <!---->
  <transition-stub name=\\"fade-in-width-expand-transition\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\"><span class=\\"n-button__icon\\" style=\\"margin: 0px;\\"><transition-stub name=\\"icon-switch-transition\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\"><div class=\\"n-icon-slot\\" role=\\"none\\"><svg viewBox=\\"0 0 32 32\\" width=\\"1.2em\\" height=\\"1.2em\\"><path fill=\\"currentColor\\" d=\\"M18 19h6v2h-6zm0-4h12v2H18zm0-4h12v2H18zm-4 10v-2H9v-2H7v2H2v2h8.215a8.591 8.591 0 0 1-2.216 3.977A9.273 9.273 0 0 1 6.552 23H4.333a10.855 10.855 0 0 0 2.145 3.297A14.658 14.658 0 0 1 3 28.127L3.702 30a16.42 16.42 0 0 0 4.29-2.336A16.488 16.488 0 0 0 12.299 30L13 28.127A14.664 14.664 0 0 1 9.523 26.3a10.313 10.313 0 0 0 2.729-5.3zm-2.833-8h2.166L8.75 2H6.583L2 13h2.166L5 11h5.333zM5.833 9l1.833-4.4L9.5 9z\\"></path></svg></div></transition-stub></span></transition-stub>
  <!---->
  <div aria-hidden=\\"true\\" class=\\"n-base-wave\\"></div>
  <div aria-hidden=\\"true\\" class=\\"n-button__border\\"></div>
  <div aria-hidden=\\"true\\" class=\\"n-button__state-border\\"></div>
</button>"
`;

exports[`LanguageSelector.vue > should render when expanded 1`] = `"<div data-v-app=\\"\\"><!----><button class=\\"n-button n-button--default-type n-button--medium-type\\" tabindex=\\"0\\" type=\\"button\\" style=\\"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-bezier-ease-out: cubic-bezier(0, 0, .2, 1); --n-ripple-duration: .6s; --n-opacity-disabled: 0.5; --n-wave-opacity: 0.6; font-weight: 400; --n-color: #0000; --n-color-hover: #0000; --n-color-pressed: #0000; --n-color-focus: #0000; --n-color-disabled: #0000; --n-ripple-color: #18a058; --n-text-color: rgb(51, 54, 57); --n-text-color-hover: #36ad6a; --n-text-color-pressed: #0c7a43; --n-text-color-focus: #36ad6a; --n-text-color-disabled: rgb(51, 54, 57); --n-border: 1px solid rgb(224, 224, 230); --n-border-hover: 1px solid #36ad6a; --n-border-pressed: 1px solid #0c7a43; --n-border-focus: 1px solid #36ad6a; --n-border-disabled: 1px solid rgb(224, 224, 230); --n-width: initial; --n-height: 34px; --n-font-size: 14px; --n-padding: 0 14px; --n-icon-size: 18px; --n-icon-margin: 6px; --n-border-radius: 3px;\\" title=\\"Change language\\" data-test=\\"trigger\\"><!----><transition-stub name=\\"fade-in-width-expand-transition\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\"><span class=\\"n-button__icon\\" style=\\"margin: 0px;\\"><transition-stub name=\\"icon-switch-transition\\" appear=\\"true\\" persisted=\\"false\\" css=\\"true\\"><div class=\\"n-icon-slot\\" role=\\"none\\"><svg viewBox=\\"0 0 32 32\\" width=\\"1.2em\\" height=\\"1.2em\\"><path fill=\\"currentColor\\" d=\\"M18 19h6v2h-6zm0-4h12v2H18zm0-4h12v2H18zm-4 10v-2H9v-2H7v2H2v2h8.215a8.591 8.591 0 0 1-2.216 3.977A9.273 9.273 0 0 1 6.552 23H4.333a10.855 10.855 0 0 0 2.145 3.297A14.658 14.658 0 0 1 3 28.127L3.702 30a16.42 16.42 0 0 0 4.29-2.336A16.488 16.488 0 0 0 12.299 30L13 28.127A14.664 14.664 0 0 1 9.523 26.3a10.313 10.313 0 0 0 2.729-5.3zm-2.833-8h2.166L8.75 2H6.583L2 13h2.166L5 11h5.333zM5.833 9l1.833-4.4L9.5 9z\\"></path></svg></div></transition-stub></span></transition-stub><!----><div aria-hidden=\\"true\\" class=\\"n-base-wave\\"></div><div aria-hidden=\\"true\\" class=\\"n-button__border\\"></div><div aria-hidden=\\"true\\" class=\\"n-button__state-border\\"></div></button><!--teleport start--><!--teleport end--></div><div class=\\"v-binder-follower-container\\"><div class=\\"v-binder-follower-content\\"><transition-stub name=\\"popover-transition\\" appear=\\"true\\" persisted=\\"false\\" css=\\"true\\"><div class=\\"n-dropdown-menu n-popover-shared n-dropdown\\" style=\\"--n-box-shadow: 0 3px 6px -4px rgba(0, 0, 0, .12), 0 6px 16px 0 rgba(0, 0, 0, .08), 0 9px 28px 8px rgba(0, 0, 0, .05); --n-bezier: cubic-bezier(.4, 0, .2, 1); --n-bezier-ease-in: cubic-bezier(.4, 0, 1, 1); --n-bezier-ease-out: cubic-bezier(0, 0, .2, 1); --n-font-size: 14px; --n-text-color: rgb(51, 54, 57); --n-color: #fff; --n-divider-color: rgb(239, 239, 245); --n-border-radius: 3px; --n-arrow-height: 6px; --n-arrow-offset: 10px; --n-arrow-offset-vertical: 10px; --n-padding: 4px 0; --n-space: 6px; --n-space-arrow: 10px; --n-option-height: 34px; --n-option-prefix-width: 14px; --n-option-icon-prefix-width: 36px; --n-option-suffix-width: 14px; --n-option-icon-suffix-width: 32px; --n-option-icon-size: 16px; --n-option-opacity-disabled: 0.5; --n-option-color-hover: rgb(243, 243, 245); --n-option-color-active: rgba(24, 160, 88, 0.1); --n-option-text-color: rgb(51, 54, 57); --n-option-text-color-hover: rgb(51, 54, 57); --n-option-text-color-active: #18a058; --n-option-text-color-child-active: #18a058; --n-prefix-color: rgb(51, 54, 57); --n-suffix-color: rgb(51, 54, 57); --n-group-header-text-color: rgb(118, 124, 130);\\"><div class=\\"n-dropdown-option\\" data-dropdown-option=\\"true\\"><div class=\\"n-dropdown-option-body\\"><div class=\\"n-dropdown-option-body__prefix n-dropdown-option-body__prefix--show-icon\\"><svg viewBox=\\"0 0 640 480\\" width=\\"1.2em\\" height=\\"1.2em\\"><path fill=\\"#ffce00\\" d=\\"M0 320h640v160H0z\\"></path><path d=\\"M0 0h640v160H0z\\"></path><path fill=\\"#d00\\" d=\\"M0 160h640v160H0z\\"></path></svg></div><div data-dropdown-option=\\"true\\" class=\\"n-dropdown-option-body__label\\">Deutsch</div><div data-dropdown-option=\\"true\\" class=\\"n-dropdown-option-body__suffix\\"></div></div><!----></div><div class=\\"n-dropdown-option\\" data-dropdown-option=\\"true\\"><div class=\\"n-dropdown-option-body n-dropdown-option-body--disabled\\"><div class=\\"n-dropdown-option-body__prefix n-dropdown-option-body__prefix--show-icon\\"><svg viewBox=\\"0 0 640 480\\" width=\\"1.2em\\" height=\\"1.2em\\"><path fill=\\"#012169\\" d=\\"M0 0h640v480H0z\\"></path><path fill=\\"#FFF\\" d=\\"m75 0l244 181L562 0h78v62L400 241l240 178v61h-80L320 301L81 480H0v-60l239-178L0 64V0h75z\\"></path><path fill=\\"#C8102E\\" d=\\"m424 281l216 159v40L369 281h55zm-184 20l6 35L54 480H0l240-179zM640 0v3L391 191l2-44L590 0h50zM0 0l239 176h-60L0 42V0z\\"></path><path fill=\\"#FFF\\" d=\\"M241 0v480h160V0H241zM0 160v160h640V160H0z\\"></path><path fill=\\"#C8102E\\" d=\\"M0 193v96h640v-96H0zM273 0v480h96V0h-96z\\"></path></svg></div><div data-dropdown-option=\\"true\\" class=\\"n-dropdown-option-body__label\\">English</div><div data-dropdown-option=\\"true\\" class=\\"n-dropdown-option-body__suffix\\"></div></div><!----></div><!----></div></transition-stub></div></div>"`;
